### ISSUE TITLE
Named init blocks priority idea

### ISSUE THREAD

**Issue by valery-iwanofu** [renpy/renpy#3565]
### Idea
For now, priority of init block takes value as int.
Theme idea, allow string names as a priority.

### Example

**0priorties.rpy**:
```renpy
python early:
    config.named_priorities['LOCATION_DEF'] = -100
```
**location_defs.rpy**:
```renpy
init LOCATION_DEF python:
    define_location(
        key='FlowerShop',
        name=_('Flower Shop'),
        parent='Mall'
    )
```

**Comment by renpytom** [comment]
It might make sense to allow these priorities to be expressions that then get evaluated sometime after the python early runs.

**Comment by Gouvernathor** [comment]
Maybe treat it like the speaker of the dialogue lines, with the `character` named store, so that the canonical way of defining constants for priorities would not leak them into the main store.

**Comment by Moshibit** [comment]
Would it be useful if the launcher shows in the Navigate Script section the init blocks ordered by priority?

**Comment by Gouvernathor** [comment]
@Moshibit if we enable Tom's idea for general expressions, that would mean executing the python early blocks of the game to be necessary before accessing the "navigate script" section. That's not the worse thing ever, but I think the section as it currently is is impervious to syntax errors and/or divisions by 0 anywhere in the project's files.

**Comment by Andykl** [comment]
I believe this is better be implemented by projects themselves with CDS that mimic `init` statement, than complicate Ren'Py itself.

**Comment by valery-iwanofu** [comment]
@Andykl Honestly, I tried to implement named init blocks via cds, but it didn't work for me, because I couldn't solve the problem with `init_priority` and just switched to using loadable modules. The problem with `init_priority` in the `renpy.register_statement` function is that the value of init_priority is static, it goes into the rpyc file and won't be updated until the source rpy file is changed. This was a problem for me because I wanted to define priorities as a simple list, where the priority value would be its index in that list.

But I think you're right that Ren'Py shouldn't be complicated for the sake of a feature that can be easily replaced by cds. Maybe instead of changing the init statement, `renpy.register_statement` should have an optional parameter get_init_priority that will be executed inside `renpy.ast.UserStatement` like other function parameters?

**Comment by Andykl** [comment]
I mean, the only difference between
0priorties.rpy:
```
python early:
    config.named_priorities['LOCATION_DEF'] = -100
```
location_defs.rpy:
```
init LOCATION_DEF python:
    define_location(
        key='FlowerShop',
        name=_('Flower Shop'),
        parent='Mall'
    )
```
and
0priorties.rpy:
```
python early:
    named_priorities = {}
    named_priorities['location'] = -100

    for name, prio in named_priorities.items():
        # implementing parse is left as an exercice
        renpy.register_statement(f"init {name}", ..., init_priority=prio)  

```
location_defs.rpy:
```
init location:
    define_location(
        key='FlowerShop',
        name=_('Flower Shop'),
        parent='Mall'
    )
```
Is that in former case your `config.named_priorities` is constraied to be before first possible init block with that name, and in later it needs to be before `register_statement`, which for me makes no difference because it always should be the code you control.

So, I don't think we need to add `get_init_priority`, for the same reason we don't provide an early config option to overwrite `image` or `screen` default init priority. But maybe Tom have a different opinion, so I keep it open.

**Comment by valery-iwanofu** [comment]
The problem is that init_priority belongs to the UserStatement instance and will be stored inside the rpyc file: https://github.com/renpy/renpy/blob/0cdb3a576913f69c4706e59f01864129ecdd066f/renpy/statements.py#L315-L321

When we change the values ​​of named_priorities, this will not affect init_priority in the already compiled rpyc files. If named_priorities is a dictionary with hardcoded numbers, you can still live with it, but if named_priorities is a list, then problems will definitely appear. If `renpy.register_statement` had a get_init_priority parameter, this problem could be easily solved:

```renpy
python early:
    named_priorities = [
        'location',
        'character'
    ]
    name_to_priority = {name: prio for name in enumerate(named_priorities)} 

    def parse_my_init(l):
        data = {}
        data['priority'] = l.require(l.name)
        ...
        return data

    def get_init_priority_of_my_init(parsed):
        return name_to_priority[parsed['priority']]
    
    renpy.register_statement(
        name="my_init",
        parse=parse_my_init,
        get_init_priority=get_init_priority_of_my_init # get_init_priority will be executed in UserStatement.get_init at runtime
    )
```

Since `get_init_priority` relies on parsed data and is not much different from `execute`, `lint`, etc., there should be no problem with early code execution.

**Comment by Andykl** [comment]
Ah, I see the problem. Yeah, I never was worried about it because we don't checkout rpyc's and enforce that Ren'Py never load rpyc's from a different commit, so I can safely change init priority of my statements. But for different workflows this could be a problem.

Changing `init_priority` from `int` to `int | Callable[[], int]` could be an option, but that is Tom's call.

**Comment by Andykl** [comment]
Hm. Alternatively, we could change `UserStatement` to always use init priority from corresponding `renpy.register_statement` and not store it on the node.

**Comment by mal** [comment]
Keep in mind that things like `pause` are `UserStatement`s, so any runtime complexity added to `get_init` would likely be reflected in boot times based on the number of pauses in a script.

**If we really need this**, my suggestion would be to allow `init_priority` to be a string, store it literally, then in [`script.py`](https://github.com/renpy/renpy/blob/0cdb3a576913f69c4706e59f01864129ecdd066f/renpy/script.py#L686) look up the str (from `get_init` in `config.named_priorities` (or whatever) only when it `is not None` before added the entry to `initcode`. That could also solve the `init BLAH:` case too, since it wouldn't be unique to `UserStatement`s.

i.e.
```py
            if (priority := node.get_init()) is not None:
                initcode.append((config.named_priorities.get(priority, priority), node))
```

We'd need to test it for speed, but short of having thousands of init blocks, it'd probably be fine.

**Final caveat**: I don't know what impact if any a mixed type has on the shiny new C-nodes, would be a shame to immediately cripple them by accident.

**Comment by valery-iwanofu** [comment]
I checked the script loading code, correct me if I'm wrong, but all UserStatements are already added to the initcode, affecting the execution time of the init phase, since UserStatement.get_init never returns None

**Comment by Andykl** [comment]
You are right. I made several fixes around CDS init logic in #6468, so it should resolve issues above.

**Comment by Andykl** [comment]
Could you check if the issue is resolved when next nightly is out?

**Comment by valery-iwanofu** [comment]
I checked if the issue was resolved using the following code:

<details>
  <summary>Code</summary>
  
0my_init_statements.rpy:

  ```renpy
python early hide:
    import functools
    
    PRIO = [
        'character',
        'location',
    ]
    PRIO_BY_NAME = {
        name: prio for prio, name in enumerate(PRIO)
    }

    def my_init_parse(lexer):
        lexer.advance()

    def my_init_execute(name, declared_order, kind, parsed):
        if PRIO_BY_NAME[name] != declared_order:
            raise Exception(f'Mismatch init order for {kind}.{name}')

    for prio, name in enumerate(PRIO):
        renpy.register_statement(
            f'my_init execute_init {name}', 
            parse=my_init_parse, 
            execute_init=functools.partial(my_init_execute, name, prio, 'execute_init'),
            init_priority=prio
        )
        renpy.register_statement(
            f'my_init execute {name}', 
            parse=my_init_parse, 
            execute_init=functools.partial(my_init_execute, name, prio, 'init'),
            init_priority=prio,
            init=True
        )

  ```

script.rpy:
```renpy
my_init execute location
my_init execute character

my_init execute_init character
my_init execute_init location
```
  
</details>

When I change the PRIO order after script.rpy is already compiled into script.rpyc, everything works as expected - init_priority is taken not from the rpyc file as before fix, but from the registered statement.

Thanks a lot for the fix. I guess the issue can be closed?
